dvc/cache/base.py

```diff
@@ -89,7 +89,23 @@ def load_dir_cache(self, hash_info):
 
         return DirInfo.from_list(d)
 
-    def changed(self, path_info, hash_info):
+    def _get_specialized_hash_info(self, path_info, hash_info, hash_origin):
+        dir_info = self.get_dir_cache(hash_info)
+        hash_key = path_info.relative_to(hash_origin).parts
+        if hash_key in dir_info.trie:
+            return path_info, dir_info.trie[hash_key]
+        else:
+            depth = len(hash_key)
+            specialized_dir_info = DirInfo()
+            try:
+                for key, value in dir_info.trie.items(hash_key):
+                    specialized_dir_info.trie[key[depth:]] = value
+            except KeyError:
+                return hash_origin, hash_info
+            else:
+                return path_info, self._get_dir_info_hash(specialized_dir_info)
```

Comment:
This method is really confusing and is doing too much. The fallback logic here doesn't really apply to the filtered hash_info (btw `filtered` is probably a better term than `specialized` here).

Also note how giant the `else` statement is after `if & return`. We don't have this check enabled in pylint yet, but we do appreciate no-else-after-return unless there is a good reason to do otherwise.